File: Starfall/Players/States/StatesManager.cs
using System;
using System.Diagnostics;

namespace Starfall.Players.States
{
    public class StatesManager
    {
        private Player _player;

        public IPlayerState CurrentPlayerState { get; private set; }
        public RunningState RunningState { get; }
        public JumpingState JumpingState { get; }

        public Stopwatch CurrentJumpTimer { get; private set; }

        private StatesManager() { }
        public StatesManager(
             Player player,
             JumpGemBar jumpGemBar)
        {
            _player = player ?? throw new ArgumentNullException(nameof(player));

            RunningState = new RunningState(player, this);
            JumpingState = new JumpingState(player, jumpGemBar, this);
            CurrentPlayerState = RunningState;
            CurrentPlayerState.Enter();
        }

        public void HandleJump()
            => CurrentPlayerState.HandleJump();

        public void Update(TimeSpan elapsed)
        {
            if (CurrentJumpTimer != null && CurrentJumpTimer.IsRunning)
                _player.SetCurrentJumpTime(CurrentJumpTimer.Elapsed);

            var newState = CurrentPlayerState.Update(elapsed);
            if (newState != CurrentPlayerState)
            {
                if (newState is JumpingState)
                {
                    CurrentJumpTimer = new Stopwatch();
                    CurrentJumpTimer.Start();
                }
                else if (CurrentJumpTimer.IsRunning)
                {
                    CurrentJumpTimer.Stop();
                }

                CurrentPlayerState = newState;
                CurrentPlayerState.Enter();
            }
        }
    }
}

File: Web/src/player/states/StatesManager.ts
import { Ticker } from "pixi.js";
import JumpGemBar from "../JumpGemBar.ts";
import IPlayerState from "./IPlayerState.ts";
import IPlayerStateContext from "./IPlayerStateContext.ts";
import JumpingState from "./JumpingState.ts";
import RunningState from "./RunningState.ts";

class StatesManager {
  private _currentState: IPlayerState;
  readonly runningState: RunningState;
  readonly jumpingState: JumpingState;

  // tracks how long the player stays airborne per jump
  private _isJumpTimerRunning = false;
  private _currentJumpDuration = 0;
  private _bestJumpDuration = 0;

  constructor(player: IPlayerStateContext, jumpGemBar: JumpGemBar) {
    this.runningState = new RunningState(player, () => this.jumpingState);
    this.jumpingState = new JumpingState(
      player,
      jumpGemBar,
      () => this.runningState,
    );

    this._currentState = this.runningState;
    this._currentState.enter();
  }

  handleJump() {
    this._currentState.handleJump();
  }

  update(time: Ticker) {
    if (this._isJumpTimerRunning) {
      this._currentJumpDuration += time.elapsedMS / 1000;
      if (this._currentJumpDuration > this._bestJumpDuration) {
        this._bestJumpDuration = this._currentJumpDuration;
      }
    }

    const newState = this._currentState.update(time);
    if (newState !== this._currentState) {
      if (newState === this.jumpingState) {
        this._isJumpTimerRunning = true;
        this._currentJumpDuration = 0;
      } else {
        this._isJumpTimerRunning = false;
      }
      this._currentState = newState;
      this._currentState.enter();
    }
  }

  get bestJumpDuration() {
    return this._bestJumpDuration;
  }
}

export default StatesManager;
